### PR TITLE
#6 Open ended question survey controller and frontend

### PR DESCRIPTION
**Pull request by MichaelDysart**

### CODE REVIEW

File: src/main/java/survey/Question.java [hunk 1]
```diff
@@ -14,6 +14,11 @@
     protected String type;
     protected String question;
 
+    /*
+     * A default constructor
+     */
+    public Question(){ }
```
Comment (TaylorAbraham) [inline src/main/java/survey/Question.java:23]: Default constructors are not needed in Java.

File: src/main/java/survey/Survey.java [hunk 2]
```diff
@@ -14,6 +14,11 @@
     @OneToMany(cascade = CascadeType.ALL)
     private Collection<Question> questions;
 
+    /*
+     * A default constructor
+     */
+    public Survey(){ }
```
Comment (TaylorAbraham) [inline src/main/java/survey/Survey.java:23]: Same thing here

File: src/main/resources/static/js/book.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    var survey;
+
+    survey = {
+        name : "survey1",
+        questions : [ { type: "openEnded", question: "q1" }, { type: "openEnded", question: "q2" } ]
+    };
+
+    $.ajax({
+         method: "POST",
+         url   : "http://localhost:8080/createSurvey",
```
Comment (TaylorAbraham) [inline src/main/resources/static/js/book.js]: We should make the URL flexible, as in have the "localhost:8080" portion defined as a constant elsewhere. I can change this when I implement React.

File: src/main/resources/templates/root.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
+    <script th:src="@{/js/book.js}"></script>
+</head>
+<body>
+<button onclick="myFunc1()">A</button>
+
+<button onclick="myFunc2()">B</button>
+
+    <div style="width:200px, height:200px, border:solid">
```
Comment (TaylorAbraham) [inline src/main/resources/templates/root.html]: We should avoid inline styles. Again, can fix in the React app.

File: src/main/java/webApp/QuestionRepository.java [hunk 5]
```diff
@@ -1,9 +1,9 @@
...
+package webApp;
 
 import org.springframework.data.repository.PagingAndSortingRepository;
 import org.springframework.data.rest.core.annotation.RepositoryRestResource;
-import survey.Question;
+import survey.OpenEndedQuestion;
 
 @RepositoryRestResource(collectionResourceRel = "survey", path = "survey")
-public interface QuestionRepository extends PagingAndSortingRepository<Question, Long> {
+public interface QuestionRepository extends PagingAndSortingRepository<OpenEndedQuestion, Long> {
```
Comment (rinadina) [inline src/main/java/webApp/QuestionRepository.java]: shouldn't the type be Question ?

File: src/main/java/survey/Question.java (hunk 1 above)
Comment (Fahid-Mannan) [inline src/main/java/survey/Question.java:23]: Comments explaining what this class does would be good

**Review by TaylorAbraham — APPROVED** [review]
Michael, reading your pull requests are like a work of art. As Shakespeare once said, "Some are born great, some achieve greatness, and some have greatness thrust upon them." [1] It would be my pleasure to add your code to our project repository.

Bibliography
[1] Shakespeare quotes, web. Accessed March 4, 2020. Available: <https://www.nosweatshakespeare.com/quotes/famous-shakespeare-quotes/>

File: src/main/java/webApp/ServingWebContentApplication.java [hunk 6]
```diff
@@ -3,12 +3,12 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.context.annotation.ComponentScan;
 
-@SpringBootApplication(scanBasePackages={"repository"})
-@EntityScan(basePackages = {"repository"})
+@SpringBootApplication
+@EntityScan(basePackages = {"survey"})
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/ServingWebContentApplication.java:11]: Can we have the word "survey" as a constant for later use in this class

File: src/main/resources/static/js/book.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
+function myFunc1() {
```
Comment (Fahid-Mannan) [inline src/main/resources/static/js/book.js]: Better name then myFunc1

**Review by Fahid-Mannan — COMMENTED** [review]
Overall good, can we have better function naming, and especially comments on the top of every class, specifying what the class does and how it fits in.

File: src/main/resources/static/js/book.js (hunk 7 above)
Comment (MichaelDysart) [reply src/main/resources/static/js/book.js]: This is a temporary function to help developers. It will be removed later.

File: src/main/java/webApp/ServingWebContentApplication.java (hunk 6 above)
Comment (MichaelDysart) [reply src/main/java/webApp/ServingWebContentApplication.java:11]: "survey" isn't used later in the class.
I don't know what the advantage of having it as a constant is. This name must correspond to a package name. Making it a constant means the reviewer has to look elsewhere to make sure that the value is correct, so doesn't save any time.

Constants make sense in situation where the value matters little but it is used in many places. For example, the ring buffer on the arduino has a macro that sets the size of the buffer and is used in the read and write functions to validate the length of the buffer to prevent overflow. There, the macro constant is used many places, but the actual value could be anything over 1, so it is easy to change. In this case, "survey" doesn't seem to be used in other places, and is also integral to the code working here so it's ability to change is limited. I suppose it might change if we rename the other package, but in that case refactoring is required anyway so unless we have it everywhere I don't know the advantage.

Maybe there is a config file where this would be more appropriate so it applies everywhere, but I don't know what that would be.

File: src/main/resources/static/js/book.js (hunk 7 above)
Comment (TaylorAbraham) [reply src/main/resources/static/js/book.js]: Let's add a todo to remove it, or name it as a developer function

File: src/main/java/survey/OpenEndedQuestion.java [hunk 8]
```diff
@@ -1,10 +1,24 @@
 package survey;
```
Comment (Fahid-Mannan) [inline src/main/java/survey/OpenEndedQuestion.java:1]: Everyone, we should implement a format where we comment on top of each class stating what the class does and its relevance to the entire project

File: src/main/java/survey/OpenEndedQuestion.java [hunk 9]
```diff
@@ -1,10 +1,24 @@
...
 import java.util.ArrayList;
 
+@Entity
 public class OpenEndedQuestion extends Question {
+
     private ArrayList<String> answers = new ArrayList<String>();
 
+    /*
+     * A default constructor
+     * @param question {String}
```
Comment (Fahid-Mannan) [inline src/main/java/survey/OpenEndedQuestion.java]: "@param question {String}" what param?

File: src/main/java/survey/OpenEndedQuestion.java [hunk 10]
```diff
@@ -32,4 +46,5 @@ public void setAnswers(ArrayList<String> answers) {
     public void addAnswer(String answer) {
```
Comment (Fahid-Mannan) [inline src/main/java/survey/OpenEndedQuestion.java:50]: I don't see any java docs here

File: src/main/java/webApp/QuestionMessage.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package webApp;
+
+public class QuestionMessage {
+
+    private final String type;
+    private final String question;
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/QuestionMessage.java:9]: Wondering if it would make any sense to have both these constants as "static"? Can someone key in an see if makes any sense

File: src/main/java/webApp/QuestionMessage.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    public QuestionMessage(String type, String question) {
+        this.type = type;
+        this.question = question;
+    }
+
+    public String getType() {
+        return type;
+    }
+
+    public String getQuestion() {
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/QuestionMessage.java:33]: Some javadocs/comments for this entire class please

File: src/main/java/webApp/ServingWebContentApplication.java [hunk 13]
```diff
@@ -3,12 +3,12 @@
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.context.annotation.ComponentScan;
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/ServingWebContentApplication.java]: what is component scan being used for?

File: src/main/java/webApp/webController.java [hunk 14]
```diff
@@ -1,56 +1,62 @@
...
-import java.util.Iterator;
-import java.util.List;
 
 @Controller
 public class webController {
     @Autowired
     private SurveyRepository repo;
 
-    @GetMapping("/index")
+    @GetMapping("/")
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/webController.java:22]: isn't the default just the index page?

File: src/main/java/webApp/webController.java [hunk 15]
```diff
@@ -1,56 +1,62 @@
...
-    public String createSurvey(@RequestParam("surveyName") String surveyName, @RequestBody Collection<Question> questions) {
-        Survey survey = new Survey(surveyName);
-        survey.setQuestions(questions);
+    @PostMapping(value = "/createSurvey", consumes = "application/json", produces = "application/json")
+    @ResponseBody
+    public Response createSurvey(@RequestBody SurveyMessage surveyMessage) {
+        Survey survey = new Survey(surveyMessage.getName());
+        Collection<Question> questionList = new ArrayList<Question>();
+
+        for (QuestionMessage question : surveyMessage.getQuestions()) {
```
Comment (Fahid-Mannan) [inline src/main/java/webApp/webController.java:33]: I am wondering if we compress this by using lambdas and filters (java 8)

Comment (Fahid-Mannan) [reply src/main/java/webApp/webController.java:33]: questionList.add(surveyMessage.getQuestions().stream().filter(question -> question.getType().equals("openEnded")).collect(Collectors.toList()));

File: src/main/java/survey/OpenEndedQuestion.java (hunk 8 above)
Comment (TaylorAbraham) [reply src/main/java/survey/OpenEndedQuestion.java:1]: Yeah I agree

File: src/main/resources/static/js/book.js (hunk 3 above)
Comment (JacksonSchoe) [reply src/main/resources/static/js/book.js]: Ya, it wouldn't be a bad idea to maybe have a constant or properties file.

File: src/main/java/webApp/ServingWebContentApplication.java (hunk 6 above)
Comment (JacksonSchoe) [reply src/main/java/webApp/ServingWebContentApplication.java:11]: So this is what referred to maybe having in the "localhost:8080" comment I had about the potential for having a properties or constant holder file. I don't know if this that would even be applicable here but we can discuss it in the future.

File: src/main/java/webApp/QuestionMessage.java (hunk 11 above)
Comment (JacksonSchoe) [reply src/main/java/webApp/QuestionMessage.java:9]: Could almost be an enum too?

File: src/main/java/survey/OpenEndedQuestion.java (hunk 8 above)
Comment (JacksonSchoe) [reply src/main/java/survey/OpenEndedQuestion.java:1]: I'm sure there is autoformatter for having fields to be inputted at the top of the class. Something like: 
" Description:
  Contributors:
  ....etc"

If not we can create our own template for classes.

File: src/main/resources/static/js/survey.js [hunk 16]
```diff
@@ -0,0 +1,108 @@
+// Test function
+// @TODO remove
+
+var weburl;
+
+$(document).ready(function () {
+    weburl = window.location.href;
+});
+
+
```
Comment (JacksonSchoe) [inline src/main/resources/static/js/survey.js]: Should have some comments to explain how stuff is working

**Review by JacksonSchoe — APPROVED** [review]
Overall very well written. No glaring issues or concerns on my end. There is some syntax stuff that we may want to discuss in the future, in particular having a properties/config file for constants and having a unified commenting template. But we need this core functionality in the code base asap, so its fine by me to merged in.
